test/e2e/features_test.go

```diff
@@ -703,3 +703,95 @@ func TestDefaultIngressClass(t *testing.T) {
 		return false
 	}, ingressWait, time.Second)
 }
+
+// TestMissingCRDsDontCrashTheController ensures that in case of missing CRDs installation in the cluster, specific
+// controllers are disabled, this fact is properly logged, and the controller doesn't crash.
+func TestMissingCRDsDontCrashTheController(t *testing.T) {
+	t.Parallel()
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	t.Log("building test cluster and environment")
+	clusterBuilder := kind.NewBuilder()
+	if clusterVersionStr != "" {
+		clusterVersion, err := semver.ParseTolerant(clusterVersionStr)
+		require.NoError(t, err)
+		clusterBuilder.WithClusterVersion(clusterVersion)
+	}
+	cluster, err := clusterBuilder.Build(ctx)
+	require.NoError(t, err)
+	var addons []clusters.Addon
+	addons = append(addons, metallb.New())
+
+	if b, err := loadimage.NewBuilder().WithImage(imageLoad); err == nil {
+		addons = append(addons, b.Build())
+	}
+
+	builder := environments.NewBuilder().WithExistingCluster(cluster).WithAddons(addons...)
+	env, err := builder.Build(ctx)
+	require.NoError(t, err)
+
+	t.Logf("building cleaner to dump diagnostics...")
+	cleaner := clusters.NewCleaner(env.Cluster())
+	defer func() {
+		if t.Failed() {
+			output, err := cleaner.DumpDiagnostics(ctx, t.Name())
+			t.Logf("%s failed, dumped diagnostics to %s", t.Name(), output)
+			assert.NoError(t, err)
+		}
+		assert.NoError(t, cleaner.Cleanup(ctx))
+		assert.NoError(t, cluster.Cleanup(ctx))
+	}()
+
+	t.Log("deploying kong components")
+	manifest, err := getTestManifest(t, dblessPath)
+	require.NoError(t, err)
+
+	manifestWithNoCRDs := stripCRDs(t, manifest)
+	deployment := deployKong(ctx, t, env, manifestWithNoCRDs)
+
+	t.Log("waiting for pod to output required logs")
+	var podName string
+	require.Eventually(t, func() bool {
+		pods, err := env.Cluster().Client().CoreV1().Pods(deployment.Namespace).List(ctx, metav1.ListOptions{
+			LabelSelector: fmt.Sprintf("app=%s", deployment.Name),
+		})
+		if err != nil || len(pods.Items) == 0 {
+			return false
+		}
+
+		podName = pods.Items[0].Name
+		logs, err := getPodLogs(ctx, t, env, deployment.Namespace, podName)
+		if err != nil {
+			return false
+		}
+
+		resources := []string{
+			"udpingresses",
+			"tcpingresses",
+			"kongingresses",
+			"ingressclassparameterses",
+			"kongplugins",
+			"kongconsumers",
+			"kongclusterplugins",
+			"ingresses",
+			"gateways",
+			"httproutes",
+		}
+		for _, resource := range resources {
+			if !strings.Contains(logs, fmt.Sprintf("Disabling the '%s' controller due to missing CRD installation", resource)) {
+				return false
+			}
+		}
+
+		return true
+	}, time.Minute, time.Second)
+
+	t.Log("waiting 2 minutes (default controller-runtime's CacheSyncTimeout)")
+	time.Sleep(2 * time.Minute)
```

Comment:
Ugh :(

Do we really, really have to make this a sleep based test? Can't we wait for a particular log line that would occur after the CacheSyncTimeout if it happened? Or perhaps we can use a different method to ensure that the controller is running fine?

Reply:
Yeah, I don't like it either. The timeout is by default 2 minutes long and any method we'll use without modifying that will require us to wait that long. I can see two different approaches:
1) use `require.Never` with 2 minute timeout (still requires us to wait the same period of time)
```	t.Log("ensuring controller didn't crash")
	require.Never(t, func() bool {
		pod, err := env.Cluster().Client().CoreV1().Pods(deployment.Namespace).Get(ctx, podName, metav1.GetOptions{})
		if err != nil {
		    return false
		}
		return containerDidntCrash(t, *pod, "ingress-controller")
	}, 2*time.Minute, time.Second*5)```
	
2) introduce env variable which will enable us to modify the timeout and make it as short as possible

Reply:
Is this timeout for sync configurable on controller runtime side? Maybe we tweak this for the duration of the test and get feedback faster?

Reply:
That's the 2nd option I posted. But we'd have to introduce config for that which kinda makes it a public interface from now on. Are we ok with that?

Reply:
I don't mean to introduce knob to tweak this outside of tests but inside them. I'm not sure how one can do it so this might not be possible. Just throwing out options here

Reply:
It's configurable in `controller.Options` which are passed to `controller.New`, like here: https://github.com/Kong/kubernetes-ingress-controller/blob/adaddeee75a4ea6c09ea223933baf7cfa70178d4/internal/controllers/gateway/gateway_controller.go#L74

I cannot find any other way to use that other than introducing such a knob and tweaking it in manifests while deploying KIC in tests. :(

Reply:
Isn't there any log-based feedback on this? If I remember correctly, the controller-runtime periodically complains if the expected crds are not installed in the cluster. Can we leverage those lines? We have a test in KGO that opens a log stream that you can check in real-time; maybe it's helpful for this purpose: https://github.com/Kong/gateway-operator/blob/9f98c7bc5bde6c341b30eb63bf0b88e32c0f97fc/test/e2e/operatorLogs_test.go#L94-L140

Reply:
Ok got it. It totally seems like a thing we could do: add a flag for this. Or even better an env var that would not have to be passed down in a dependency injection manner.

Reply:
I've added `TEST_KONG_CONTROLLERS_CACHE_SYNC_TIMEOUT` flag for that.